src/components/tl/mlx5/mcast/tl_mlx5_mcast_one_sided_reliability.c

```diff
@@ -0,0 +1,213 @@
+{
+    ucc_tl_mlx5_team_t            *tl_team  = ucc_derived_of(team, ucc_tl_mlx5_team_t);
+    ucc_status_t                   status   = UCC_OK;
+    ucc_tl_mlx5_mcast_coll_comm_t *comm     = tl_team->mcast->mcast_comm;
+    int                            one_sided_total_slots_size, i;
+
+    /* this array keeps track of the number of recv packets from each process
+     * used in all the protocols */
+    comm->recv_list = ucc_calloc(1, comm->commsize * sizeof(uint32_t), "recv_list");
+    if (!comm->recv_list) {
+        tl_error(comm->lib, "unable to malloc for recv_list");
+        status = UCC_ERR_NO_MEMORY;
+        goto failed;
+    }
+
+    /* this array holds all the remote-addr/rkey of sendbuf from processes in the team 
+     * used in sync design. Needs to be set during each mcast-allgather call after sendbuf 
+     * registeration */
+    comm->slot_mem_info_list = ucc_calloc(1, comm->commsize * sizeof(ucc_tl_mlx5_mcast_slot_mem_info_t),
+                                          "slot_mem_info_list");
+    if (!comm->slot_mem_info_list) {
+        tl_error(comm->lib, "unable to malloc for slot_mem_info_list");
+        status = UCC_ERR_NO_MEMORY;
+        goto failed;
+    }
+
+    /* below data structures are used in async design only */
+    comm->one_sided_async_slots_size = comm->one_sided_reliability_scheme_msg_threshold
+                                           + ONE_SIDED_SLOTS_INFO_SIZE;
+    one_sided_total_slots_size       = comm->one_sided_async_slots_size *
+                                            ONE_SIDED_SLOTS_COUNT * sizeof(char);
+    comm->one_sided_slots_buffer     = (char *)ucc_calloc(1, one_sided_total_slots_size,
+                                                          "one_sided_slots_buffer");
+    if (!comm->one_sided_slots_buffer) {
+        tl_error(comm->lib, "unable to malloc for one_sided_slots_buffer");
+        status = UCC_ERR_NO_MEMORY;
+        goto failed;
+    }
+    comm->one_sided_slots_mr = ibv_reg_mr(comm->ctx->pd, comm->one_sided_slots_buffer,
+                                          one_sided_total_slots_size, IBV_ACCESS_LOCAL_WRITE |
+                                          IBV_ACCESS_REMOTE_READ | IBV_ACCESS_REMOTE_WRITE);
+    if (!comm->one_sided_slots_mr) {
+        tl_error(comm->lib, "unable to register for one_sided_slots_mr");
+        status = UCC_ERR_NO_RESOURCE;
+        goto failed;
+    }
+    
+    /* this array holds local information about the slot status that was read from remote ranks */
+    comm->remote_slot_info = ucc_calloc(1, comm->commsize * ONE_SIDED_SLOTS_INFO_SIZE,
+                                        "remote_slot_info");
+    if (!comm->remote_slot_info) {
+        tl_error(comm->lib, "unable to malloc for remote_slot_info");
+        status = UCC_ERR_NO_MEMORY;
+        goto failed;
+    }
+    comm->remote_slot_info_mr = ibv_reg_mr(comm->ctx->pd, comm->remote_slot_info,
```

Comment:
If this `ibv_reg_mr` fails, does that mean the `ibv_reg_mr` for `comm->one_sided_slots_mr` needs to be dereg'd?